### PR TITLE
Add diagnostic for dictionary

### PR DESCRIPTION
**Pull request by kimdv**
Depends on #1449 and #1382

### CODE REVIEW

File: Tests/SwiftParserTest/translated/RecoveryTests.swift [hunk 1]
```diff
@@ -810,6 +810,7 @@ final class RecoveryTests: XCTestCase {
   }
 
   func testRecovery63() {
+
```
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/RecoveryTests.swift:813]: Superfluous newline?

File: Sources/SwiftParser/Types.swift [hunk 2]
```diff
@@ -1069,6 +1069,20 @@ extension Parser {
             arena: self.arena
           )
         )
+      } else if let colon = self.consume(if: .colon) {
```
Comment (ahoppen) [inline Sources/SwiftParser/Types.swift:1072]: I’m a little worried that we enter into this path and give a weird diagnostic if somebody tries to do a Python-style function declaration like the following. It would probably be better to only recover to a dictionary type if we have colon followed by a type. And yes, I know that this doesn’t match the behavior of the old parser.

```swift
func foo() -> Int:
  // some Code
```

File: Tests/SwiftParserTest/translated/RecoveryTests.swift [hunk 3]
```diff
@@ -831,7 +832,8 @@ final class RecoveryTests: XCTestCase {
           message: "expected ':' in type annotation",
           fixIts: ["insert ':'"]
         ),
-        DiagnosticSpec(locationMarker: "3️⃣", message: #"unexpected code ': Int = ""' before function"#),
+        DiagnosticSpec(
+            locationMarker: "3️⃣", message: #"unexpected code ': Int = ""' before function"#),
```
Comment (ahoppen) [inline Tests/SwiftParserTest/translated/RecoveryTests.swift:836]: Unrelated change?

File: Sources/SwiftParser/Types.swift (hunk 2 above)
Comment (kimdv) [reply Sources/SwiftParser/Types.swift]: Something like this. 
Also added a test case for the scenario above.

**Comment by kimdv** [comment]
Depends on #1580

File: Sources/SwiftParser/Types.swift [hunk 4]
```diff
@@ -1069,6 +1069,30 @@ extension Parser {
             arena: self.arena
           )
         )
+      } else if self.at(.colon) {
+        var lookahead = self.lookahead()
+
+        // We only want to continue with a dictionary if we can parse a colon and a simpletype.
+        // Otherwise we can get a wrong diagnostiv if we get a Python-style function declaration.
```
Comment (ahoppen) [inline Sources/SwiftParser/Types.swift:1076]:
```suggestion
        // Otherwise we can get a wrong diagnostic if we get a Python-style function declaration.
```

**Review by ahoppen — APPROVED** [review]

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test

**Comment by kimdv** [comment]
@swift-ci please test linux

**Review by ahoppen — APPROVED** [review]

**Comment by kimdv** [comment]
@swift-ci please test linux

**Comment by kimdv** [comment]
@swift-ci please test Windows
